lib/iris/cube.py

```diff
@@ -3379,10 +3379,15 @@ def aggregated_by(self, coords, aggregator, **kwargs):
         # coordinate dimension.
         shared_coords = list(filter(
             lambda coord_: coord_ not in groupby_coords,
-            self.coords(dimensions=dimension_to_groupby)))
+            self.coords(contains_dimension=dimension_to_groupby)))
+
+        match_dims = [
+            index for coord_ in shared_coords for (index, dim) in
+            enumerate(self.coord_dims(coord_)) if dim == dimension_to_groupby]
```

Comment:
At the very least, instead of having to `zip(shared_coords, match_dims)`, you can have this comprehension return both of them right off the bat, no?
```
        match_dims = [  # maybe a new name...
            (coord_, index)
            for coord_ in shared_coords
            for (index, dim) in enumerate(self.coord_dims(coord_))
            if dim == dimension_to_groupby]
```

Reply:
I couldn't get my head around how to deconstruct this:  I just ended up with loops that said the same thing in a different order, which was no less confusing!  So I have gone with @QuLogic's suggestion, with an extra comment so it's at least obvious what I'm aiming for.